File: crates/subterra/src/gfx.rs
//! Graphics-related representations.

use std::io::Cursor;

use byteorder::{ByteOrder, LittleEndian, ReadBytesExt};
use util::{io::CursorExt, Id8};

use crate::Error;

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, bytemuck::Zeroable, bytemuck::Pod)]
pub struct Rgb8 {
	pub r: u8,
	pub g: u8,
	pub b: u8,
}

/// See <https://doomwiki.org/wiki/COLORMAP> (and [`ColorMapSet`]).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ColorMap(pub [u8; 256]);

impl std::ops::Deref for ColorMap {
	type Target = [u8; 256];

	fn deref(&self) -> &Self::Target {
		&self.0
	}
}

impl std::ops::DerefMut for ColorMap {
	fn deref_mut(&mut self) -> &mut Self::Target {
		&mut self.0
	}
}

/// See <https://doomwiki.org/wiki/COLORMAP>.
///
/// This type is meant to resemble a [`std::borrow::Cow`], since its content
/// can be cheaply cast directly from raw bytes and used thus.
#[derive(Debug, Clone)]
pub enum ColorMapSet<'b> {
	Borrowed(&'b [ColorMap; 34]),
	Owned(Box<[ColorMap; 34]>),
}

impl ColorMapSet<'_> {
	/// This returns a [`Self::Borrowed`] by casting the bytes directly,
	/// so it is allocation-free and very cheap.
	pub fn new(bytes: &[u8]) -> Result<Self, Error> {
		#[repr(C)]
		#[derive(Clone, Copy, bytemuck::AnyBitPattern)]
		struct Raw([[u8; 256]; 34]);

		if bytes.len() < (256 * 34) {
			return Err(Error::SizeMismatch {
				expected: 256 * 34,
				actual: bytes.len(),
			});
		}

		let raw = bytemuck::from_bytes::<Raw>(bytes);

		// SAFETY: `Raw` and `[ColorMap; 34]` have identical representations.
		unsafe { Ok(Self::Borrowed(std::mem::transmute(raw))) }
	}

	#[must_use]
	pub fn into_owned(self) -> [ColorMap; 34] {
		match self {
			Self::Borrowed(b) => b.to_owned(),
			Self::Owned(o) => *o,
		}
	}
}

impl std::ops::Index<usize> for ColorMapSet<'_> {
	type Output = ColorMap;

	fn index(&self, index: usize) -> &Self::Output {
		match self {
			Self::Borrowed(r) => &r[index],
			Self::Owned(b) => &b[index],
		}
	}
}

/// See <https://doomwiki.org/wiki/ENDOOM>.
#[derive(Debug, Clone)]
pub struct EnDoom {
	pub colors: Box<[u8; 2000]>,
	pub text: Box<[u8; 2000]>,
}

impl EnDoom {
	pub fn new(bytes: &[u8]) -> Result<Self, Error> {
		if bytes.len() != 4000 {
			return Err(Error::SizeMismatch {
				expected: 4000,
				actual: bytes.len(),
			});
		}

		let mut ret = Self {
			colors: Box::new([0; 2000]),
			text: Box::new([0; 2000]),
		};

		let mut r_i = 0;
		let mut b_i = 0;

		while b_i < 4000 {
			ret.colors[r_i] = bytes[b_i];
			ret.text[r_i] = bytes[b_i + 1];
			r_i += 1;
			b_i += 2;
		}

		Ok(ret)
	}

	#[must_use]
	pub fn is_blinking(&self, index: usize) -> bool {
		assert!(index < 2000);
		self.colors[index] & (1 << 7) == (1 << 7)
	}
}

/// See <https://doomwiki.org/wiki/PLAYPAL> (and [`PaletteSet`]).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Palette(pub [Rgb8; 256]);

impl std::ops::Deref for Palette {
	type Target = [Rgb8; 256];

	fn deref(&self) -> &Self::Target {
		&self.0
	}
}

impl std::ops::DerefMut for Palette {
	fn deref_mut(&mut self) -> &mut Self::Target {
		&mut self.0
	}
}

/// See <https://doomwiki.org/wiki/PLAYPAL>.
///
/// This type is meant to resemble a [`std::borrow::Cow`], since its content
/// can be cheaply cast directly from raw bytes and used thus.
#[derive(Debug)]
pub enum PaletteSet<'b> {
	Borrowed(&'b [Palette; 14]),
	Owned(Box<[Palette; 14]>),
}

impl PaletteSet<'_> {
	/// This returns a [`Self::Borrowed`] by casting the bytes directly to
	/// [`Palette`]s, so it is allocation-free and very cheap.
	pub fn new(bytes: &[u8]) -> Result<Self, Error> {
		#[repr(C)]
		#[derive(Clone, Copy, bytemuck::AnyBitPattern)]
		struct Raw([[[u8; 3]; 256]; 14]);

		if bytes.len() < (256 * 3 * 14) {
			return Err(Error::SizeMismatch {
				expected: 256 * 3 * 14,
				actual: bytes.len(),
			});
		}

		let raw = bytemuck::from_bytes::<Raw>(bytes);

		// SAFETY: `Raw` and `[Palette; 14]` have identical representations.
		unsafe { Ok(Self::Borrowed(std::mem::transmute(raw))) }
	}

	#[must_use]
	pub fn into_owned(self) -> [Palette; 14] {
		match self {
			Self::Borrowed(b) => b.to_owned(),
			Self::Owned(o) => *o,
		}
	}
}

impl std::ops::Index<usize> for PaletteSet<'_> {
	type Output = Palette;

	fn index(&self, index: usize) -> &Self::Output {
		match self {
			Self::Borrowed(r) => &r[index],
			Self::Owned(b) => &b[index],
		}
	}
}

/// See <https://doomwiki.org/wiki/PNAMES>.
#[derive(Debug, Default)]
pub struct PatchTable(pub Vec<Id8>);

impl std::ops::Deref for PatchTable {
	type Target = Vec<Id8>;

	fn deref(&self) -> &Self::Target {
		&self.0
	}
}

impl std::ops::DerefMut for PatchTable {
	fn deref_mut(&mut self) -> &mut Self::Target {
		&mut self.0
	}
}

impl PatchTable {
	/// Returns `Ok(None)` if the given PNAMES lump is valid,
	/// but reports itself to have 0 records in it.
	pub fn new(bytes: &[u8]) -> Result<Option<Self>, Error> {
		const RECORD_SIZE: usize = 8;

		let mut invalid = false;

		invalid |= bytes.len() < RECORD_SIZE;
		invalid |= ((bytes.len() - 4) % RECORD_SIZE) != 0;

		let len = LittleEndian::read_u32(bytes) as usize;

		if len == 0 {
			return Ok(None);
		}

		let expected = (len * RECORD_SIZE) + 4;
		invalid |= bytes.len() != expected;

		if invalid {
			return Err(Error::SizeMismatch {
				expected,
				actual: bytes.len(),
			});
		}

		let mut ret = Vec::with_capacity(len);
		let mut pos = 4;

		while pos < bytes.len() {
			let raw = bytemuck::from_bytes::<[u8; RECORD_SIZE]>(&bytes[pos..(pos + RECORD_SIZE)]);

			if let Some(pname) = util::read_id8(*raw) {
				ret.push(pname);
			}

			pos += RECORD_SIZE;
		}

		Ok(Some(Self(ret)))
	}
}

/// See <https://doomwiki.org/wiki/Picture_format>.
///
/// Partially adapted from SLADE's `DoomGfxDataFormat::isThisFormat`.
#[derive(Debug)]
pub struct PictureReader<'a> {
	bytes: &'a [u8],
	/// Short for "header cursor".
	cursor_h: Cursor<&'a [u8]>,
	/// The position just past the header.
	checkpoint: u64,
	width: u16,
	height: u16,
	left: i16,
	top: i16,
}

impl<'a> PictureReader<'a> {
	const HEADER_SIZE: usize = std::mem::size_of::<u16>() * 4;

	/// Ensure that `bytes` is the entire lump.
	/// This does not allocate, so `PictureReader::new.is_ok()` is a suitable
	/// way to check if a WAD entry is a picture-format graphic.
	pub fn new(bytes: &'a [u8]) -> Result<PictureReader<'a>, Error> {
		if bytes.len() < Self::HEADER_SIZE {
			return Err(Error::MissingHeader {
				expected: Self::HEADER_SIZE,
			});
		}

		let mut cursor_h = Cursor::new(bytes);

		let width = cursor_h.read_u16::<LittleEndian>().unwrap();
		let height = cursor_h.read_u16::<LittleEndian>().unwrap();
		let left = cursor_h.read_i16::<LittleEndian>().unwrap();
		let top = cursor_h.read_i16::<LittleEndian>().unwrap();

		// (SLADE) Sanity checks on dimensions and offsets.

		if width >= 4096 || height >= 4096 {
			return Err(Error::InvalidHeader {
				details: "width or height is >= 4096",
			});
		}

		if left <= -2000 || left >= 2000 {
			return Err(Error::InvalidHeader {
				details: "left <= -2000 or >= 2000",
			});
		}

		if top <= -2000 || top >= 2000 {
			return Err(Error::InvalidHeader {
				details: "top <= -2000 or >= 2000",
			});
		}

		if bytes.len() < (Self::HEADER_SIZE + (width as usize * 4)) {
			return Err(Error::InvalidHeader {
				details: "lump length < (header size + width)",
			});
		}

		let checkpoint = cursor_h.position(); // Just after the header.

		for _ in 0..width {
			let col_offs = cursor_h.read_u32::<LittleEndian>().unwrap() as usize;

			if col_offs > bytes.len() || col_offs < (Self::HEADER_SIZE) {
				return Err(Error::InvalidHeader {
					details: "column offset > lump length OR column offset < header size",
				});
			}

			// (SLADE) Check if total size is reasonable; this computation corresponds
			// to the most inefficient possible use of space by the format
			// (horizontal stripes of 1 pixel, 1 pixel apart).
			let num_pixels = ((height + 2 + height % 2) / 2) as usize;
			let max_col_size = std::mem::size_of::<u32>() + (num_pixels * 5) + 1;

			if bytes.len() > Self::HEADER_SIZE + (width as usize * max_col_size) {
				// Q: Unlikely, but possible. Should we try?
				return Err(Error::InvalidHeader {
					details: "lump length > (header size + (width times maximum column size))",
				});
			}
		}

		Ok(Self {
			bytes,
			cursor_h,
			checkpoint,
			width,
			height,
			left,
			top,
		})
	}

	#[must_use]
	pub fn width(&self) -> u16 {
		self.width
	}

	#[must_use]
	pub fn height(&self) -> u16 {
		self.height
	}

	/// The first element is the offset in pixels to the left of the origin;
	/// the second element is the offset below the origin.
	#[must_use]
	pub fn offset(&self) -> (i16, i16) {
		(self.left, self.top)
	}

	/// `callback`'s first two parameters are a row and column index respectively.
	pub fn read<F>(mut self, palette: &Palette, colormap: &ColorMap, mut callback: F)
	where
		F: FnMut(u32, u32, Rgb8),
	{
		let mut cursor_pix = Cursor::new(self.bytes);
		self.cursor_h.set_position(self.checkpoint);

		for i in 0..self.width {
			let col_offs = self.cursor_h.read_u32::<LittleEndian>().unwrap() as u64;
			cursor_pix.set_position(col_offs);
			let mut row_start = 0;

			while row_start != 255 {
				row_start = cursor_pix.read_u8().unwrap();

				if row_start == 255 {
					break;
				}

				let pixel_count = cursor_pix.read_u8().unwrap();
				let _ = cursor_pix.read_u8().unwrap(); // Dummy

				for ii in 0..(pixel_count as usize) {
					let map_entry = cursor_pix.read_u8().unwrap();
					let pal_entry = colormap[map_entry as usize];
					let pixel = palette[pal_entry as usize];
					let row = i as u32;
					let col = (ii as u32) + (row_start as u32);
					callback(row, col, pixel);
				}

				let _ = cursor_pix.read_u8().unwrap(); // Dummy
			}
		}
	}
}

/// See <https://doomwiki.org/wiki/TEXTURE1_and_TEXTURE2>.
#[derive(Debug, Default)]
pub struct TextureX(pub Vec<PatchedTex>);

impl std::ops::Deref for TextureX {
	type Target = Vec<PatchedTex>;

	fn deref(&self) -> &Self::Target {
		&self.0
	}
}

impl std::ops::DerefMut for TextureX {
	fn deref_mut(&mut self) -> &mut Self::Target {
		&mut self.0
	}
}

/// See [`TextureX`].
#[derive(Debug)]
pub struct PatchedTex {
	pub name: Id8,
	pub size_x: u32,
	pub size_y: u32,
	pub patches: Vec<TexPatch>,
}

/// See [`PatchedTex`].
#[derive(Debug)]
pub struct TexPatch {
	/// X-offset of this patch relative to the upper-left of the whole texture.
	pub origin_x: i32,
	/// Y-offset of this patch relative to the upper-left of the whole texture.
	pub origin_y: i32,
	/// Index into [`PatchTable`].
	pub index: usize,
}

impl TextureX {
	pub fn new(bytes: &[u8]) -> Result<Option<Self>, Error> {
		#[repr(C)]
		#[derive(Debug, Clone, Copy, PartialEq, Eq, bytemuck::AnyBitPattern)]
		struct RawMapTexture {
			name: [u8; 8],
			/// C boolean, unused.
			masked: i32,
			width: i16,
			height: i16,
			/// Unused.
			col_dir: i32,
			patch_count: i16,
		}

		#[repr(C)]
		#[derive(Debug, Clone, Copy, PartialEq, Eq, bytemuck::AnyBitPattern)]
		struct RawMapPatch {
			origin_x: i16,
			origin_y: i16,
			/// Index into [`PatchTable`].
			patch: i16,
			/// Unused.
			stepdir: i16,
			/// Unused.
			colormap: i16,
		}

		if bytes.len() < 4 {
			return Err(Error::MissingHeader { expected: 4 });
		}

		let num_textures = LittleEndian::read_u32(bytes) as usize;

		if num_textures == 0 {
			return Ok(None);
		}

		let mut curs_maptex = Cursor::new(bytes);
		curs_maptex.set_position(curs_maptex.position() + 4);

		let mut ret = Vec::with_capacity(num_textures);

		for _ in 0..num_textures {
			let start = curs_maptex.read_i32::<LittleEndian>().unwrap() as usize;
			let end = start + 22;

			if end > bytes.len() {
				return Err(Error::MissingRecord {
					expected: end,
					actual: bytes.len(),
				});
			}

			let mut curs_patch = Cursor::new(bytes);
			curs_patch.set_position(start as u64);

			let raw_tex = RawMapTexture {
				name: *curs_patch.read_from_bytes::<[u8; 8]>(),
				masked: curs_patch.read_i32::<LittleEndian>().unwrap(),
				width: curs_patch.read_i16::<LittleEndian>().unwrap(),
				height: curs_patch.read_i16::<LittleEndian>().unwrap(),
				col_dir: curs_patch.read_i32::<LittleEndian>().unwrap(),
				patch_count: curs_patch.read_i16::<LittleEndian>().unwrap(),
			};

			let patch_count = raw_tex.patch_count as usize;

			debug_assert_eq!(curs_patch.position(), end as u64);

			let mut patches = Vec::with_capacity(patch_count);

			for _ in 0..patch_count {
				let end = curs_patch.position() + (std::mem::size_of::<RawMapPatch>() as u64);

				if end as usize > bytes.len() {
					return Err(Error::MissingRecord {
						expected: end as usize,
						actual: bytes.len(),
					});
				}

				let range = (curs_patch.position() as usize)..(end as usize);
				let raw_patch = bytemuck::from_bytes::<RawMapPatch>(&bytes[range]);

				patches.push(TexPatch {
					origin_x: raw_patch.origin_x as i32,
					origin_y: raw_patch.origin_y as i32,
					index: raw_patch.patch as usize,
				});

				curs_patch.set_position(end);
			}

			ret.push(PatchedTex {
				name: util::read_id8(raw_tex.name).unwrap_or_default(),
				size_x: raw_tex.width as u32,
				size_y: raw_tex.height as u32,
				patches,
			});
		}

		Ok(Some(Self(ret)))
	}
}

#[cfg(test)]
mod test {
	use super::*;

	#[test]
	fn picture_reader() {
		let pic = include_bytes!("../../sample/freedoom/STFST01.lmp");
		let reader = PictureReader::new(pic).unwrap();
		assert_eq!(reader.width(), 24);
		assert_eq!(reader.height(), 29);
	}
}

File: libs/subterra/src/gfx.zig
//! Graphics-related representations.

const std = @import("std");

/// See <https://doomwiki.org/wiki/COLORMAP>.
/// This struct is meant to be cast to via `std.mem.bytesAsValue`.
pub const ColorMap = extern struct {
    pub const Table = [256]u8;

    /// 0 is brightest, 31 is darkest.
    brightness: [32]Table,
    invuln: Table,
    allblack: Table,
};

/// See <https://doomwiki.org/wiki/ENDOOM>.
/// This struct is meant to be cast to via `std.mem.bytesAsValue`.
pub const EnDoom = extern struct {
    pub const Foreground = enum(u4) {
        black,
        blue,
        green,
        cyan,
        red,
        magenta,
        brown,
        light_gray,
        dark_gray,
        light_blue,
        light_green,
        light_cyan,
        light_red,
        light_magenta,
        yellow,
        white,
    };

    pub const Background = enum(u3) {
        black,
        blue,
        green,
        cyan,
        red,
        magenta,
        brown,
        light_gray,
    };

    pub const Character = struct {
        ascii: u8,
        color: packed struct(u8) {
            fg: Foreground,
            bg: Background,
            blink: bool,
        },
    };

    chars: [80 * 25]Character,
};

/// See <https://doomwiki.org/wiki/PLAYPAL> and [`PaletteSet`].
pub const Palette = [256]Rgb8;

/// See <https://doomwiki.org/wiki/PLAYPAL>.
/// This struct is meant to be cast to via `std.mem.bytesAsValue`.
pub const PaletteSet = struct {
    normal: Palette,
    /// Went unused by the original Doom engine.
    red11: Palette,
    reds: [7]Palette,
    /// Went unused by the original Doom engine.
    yellow12_5: Palette,
    yellows: [3]Palette,
    green: Palette,
};

pub const PictureReader = struct {
    pub const ReadContext = fn (usize, usize, Rgb8) void;

    fbs_h: std.io.FixedBufferStream([]const u8),

    /// The position just past the header.
    checkpoint: u64,
    width: u16,
    height: u16,
    /// Offset in pixels to the left of the origin.
    left: i16,
    /// Offset below the origin.
    top: i16,

    const header_size: usize = @sizeOf(u16) * 4;

    pub fn init(bytes: []const u8) !PictureReader {
        var fbs_h = std.io.fixedBufferStream(bytes);

        const width = try fbs_h.reader().readInt(u16, .little);
        const height = try fbs_h.reader().readInt(u16, .little);
        const left = try fbs_h.reader().readInt(i16, .little);
        const top = try fbs_h.reader().readInt(i16, .little);

        // (SLADE) Sanity checks on dimensions and offsets.

        if (width >= 4096 or height >= 4096) {
            // TODO: add details to these errors when doing so is supported.
            return error.InvalidHeader;
        }

        if (left <= -2000 or left >= 2000) {
            return error.InvalidHeader;
        }

        if (top <= -2000 or top >= 2000) {
            return error.InvalidHeader;
        }

        if (bytes.len < (header_size + (width * 4))) {
            return error.InvalidHeader;
        }

        const checkpoint = fbs_h.pos;

        for (0..width) |_| {
            const col_offs = try fbs_h.reader().readInt(u32, .little);

            if (col_offs > bytes.len or col_offs < header_size) {
                return error.InvalidHeader;
            }

            // (SLADE) Check if total size is reasonable; this computation corresponds
            // to the most inefficient possible use of space by the format
            // (horizontal stripes of 1 pixel, 1 pixel apart).

            const num_pixels: usize = ((height + 2 + @rem(height, 2)) / 2);
            const max_col_size = @sizeOf(u32) + (num_pixels * 5) + 1;

            if (bytes.len > header_size + (width * max_col_size)) {
                // TODO: consider giving this a try, since it's unlikely but possible.
                return error.InvalidHeader;
            }
        }

        return PictureReader{
            .fbs_h = fbs_h,
            .checkpoint = checkpoint,
            .width = width,
            .height = height,
            .left = left,
            .top = top,
        };
    }

    /// `ctx` should be a type with a method `callback`, matching the signature of `ReadContext`.
    pub fn readAll(
        self: *PictureReader,
        palette: *const Palette,
        colormap: *const ColorMap.Table,
        ctx: anytype,
    ) void {
        const fbs_pix = std.io.fixedBufferStream(self.fbs_h.buffer);
        self.fbs_h.seekTo(self.fbs_h.getPos() catch unreachable) catch unreachable;

        for (0..self.width) |i| {
            const col_offs = self.fbs_h.reader().readInt(u32, .little) catch unreachable;
            fbs_pix.seekTo(col_offs) catch unreachable;
            var row_start: u8 = 0;

            while (row_start != 255) {
                row_start = fbs_pix.reader().readByte() catch unreachable;

                if (row_start == 255) break;

                const pixel_count = fbs_pix.reader().readByte() catch unreachable;
                _ = fbs_pix.reader().readByte() catch unreachable; // Dummy

                for (0..pixel_count) |ii| {
                    const map_entry = fbs_pix.reader().readByte() catch unreachable;
                    const pal_entry = colormap[map_entry];
                    const pixel = palette[pal_entry];
                    const col = ii + row_start;
                    ctx.callback(i, col, pixel);
                }

                _ = fbs_pix.reader().readByte() catch unreachable; // Dummy
            }
        }
    }
};

/// A single pixel.
pub const Rgb8 = struct {
    r: u8,
    g: u8,
    b: u8,
};

pub const PatchedTex = struct {
    /// Prefer calling `name` to reading this since it properly handles
    /// the intermediate null terminator which may or may not be present.
    name_raw: [8:0]u8,
    size_x: u32,
    size_y: u32,

    pub fn name(self: *const PatchedTex) []const u8 {
        return std.mem.sliceTo(self.name_raw[0..], 0);
    }
};

/// See <https://doomwiki.org/wiki/TEXTURE1_and_TEXTURE2>.
pub const TextureX = struct {
    fbs: std.io.FixedBufferStream([]align(@alignOf(u32)) const u8),
    num_textures: u32,
    offsets: []const u32,
    /// Index into `offsets`.
    pos: usize,

    pub fn init(bytes: []align(@alignOf(u32)) const u8) !TextureX {
        if (bytes.len < @sizeOf(u32)) {
            return error.InvalidHeader;
        }

        var fbs = std.io.fixedBufferStream(bytes);

        const num_textures = try fbs.reader().readInt(u32, .little);

        const offsets_start = @sizeOf(u32);
        const offsets_end = offsets_start + (@sizeOf(u32) * num_textures);

        if (bytes.len < offsets_end) {
            // Not enough bytes left for the expected offsets array.
            return error.InvalidHeader;
        }

        const offsets = std.mem.bytesAsSlice(u32, bytes[offsets_start..offsets_end]);
        const expected_len = offsets[offsets.len - 1] + @sizeOf(MapTexture);

        if (bytes.len < expected_len) {
            return error.InvalidHeader;
        }

        return TextureX{
            .fbs = fbs,
            .num_textures = num_textures,
            .offsets = offsets,
            .pos = 0,
        };
    }

    pub fn reset(self: *TextureX) void {
        self.pos = 0;
        self.fbs.seekTo(@sizeOf(u32) + (@sizeOf(u32) * self.num_textures)) catch unreachable;
    }

    pub fn next(self: *TextureX) !?MapPatches {
        if (self.pos >= self.offsets.len) {
            return null;
        }

        self.fbs.seekTo(self.offsets[self.pos]) catch unreachable;

        const texture = self.fbs.reader().readStructEndian(MapTexture, .little) catch unreachable;
        const expected_bytes = self.fbs.pos + (@sizeOf(TexPatch) * texture.patch_count);

        if (self.fbs.buffer.len < expected_bytes) {
            return error.EndOfStream;
        }

        const patch_bytes: []align(@alignOf(TexPatch)) const u8 = @alignCast(
            self.fbs.buffer[self.fbs.pos..expected_bytes],
        );

        const patches = std.mem.bytesAsSlice(TexPatch, patch_bytes);

        self.pos += 1;
        return MapPatches{ .texture = texture, .patches = patches };
    }
};

pub const MapPatches = struct {
    texture: MapTexture,
    patches: []align(@alignOf(TexPatch)) const TexPatch,
};

/// See <https://doomwiki.org/wiki/TEXTURE1_and_TEXTURE2> and [`TextureX`].
pub const MapTexture = extern struct {
    /// Prefer calling `name` to reading this since it properly handles
    /// the intermediate null terminator which may or may not be present.
    name_raw: [8]u8 align(1),
    /// Acts like a boolean (0 = false, 1 = true).
    masked: u32 align(2),
    width: u16,
    height: u16,
    /// Unused by the original Doom engine.
    column_directory: u32 align(2),
    patch_count: u16,

    pub fn name(self: *const MapTexture) []const u8 {
        return std.mem.sliceTo(self.name_raw[0..], 0);
    }
};

/// See [`PatchedTex`].
/// This struct is meant to be cast to via `std.mem.bytesAsValue`.
pub const TexPatch = extern struct {
    /// X-offset of this patch relative to the upper-left of the whole texture.
    origin_x: i16,
    /// Y-offset of this patch relative to the upper-left of the whole texture.
    origin_y: i16,
    /// Index into [`PatchTable`].
    index: u16,
    /// Unused by the original Doom engine.
    step_dir: i16,
    /// Unused by the original Doom engine.
    colormap: i16,
};

// TODO: would be nice if sample data could be embedded in the test binary
// without having to put the files in this directory.

test "PictureReader smoke" {
    const f = try std.fs.cwd().openFile("sample/freedoom/STFST01.lmp", .{});
    defer f.close();

    const bytes = try f.readToEndAlloc(std.testing.allocator, 1024);
    defer std.testing.allocator.free(bytes);

    const pic = try PictureReader.init(bytes);
    try std.testing.expectEqual(pic.width, 24);
    try std.testing.expectEqual(pic.height, 29);
}

test "TEXTUREx smoke" {
    const f = try std.fs.cwd().openFile("sample/freedoom/TEXTURE2.lmp", .{});
    defer f.close();

    const bytes = try f.readToEndAllocOptions(
        std.testing.allocator,
        1024 * 8,
        null,
        @alignOf(u32),
        null,
    );
    defer std.testing.allocator.free(bytes);

    var texx = try TextureX.init(bytes);

    _ = try texx.next() orelse unreachable;
    _ = try texx.next() orelse unreachable;

    const bigdoor5 = try texx.next() orelse unreachable;

    try std.testing.expectEqualStrings("BIGDOOR5", bigdoor5.texture.name());
    try std.testing.expectEqual(bigdoor5.texture.patch_count, 4);
    try std.testing.expectEqual(128, bigdoor5.texture.height);
    try std.testing.expectEqual(128, bigdoor5.texture.width);
    try std.testing.expectEqual(4, bigdoor5.patches.len);

    try std.testing.expectEqual(
        [2]i16{ 0, 0 },
        [2]i16{ bigdoor5.patches[0].origin_x, bigdoor5.patches[0].origin_y },
    );
    try std.testing.expectEqual(
        [2]i16{ 104, 0 },
        [2]i16{ bigdoor5.patches[3].origin_x, bigdoor5.patches[3].origin_y },
    );

    while (try texx.next()) |_| {}
    texx.reset();
}
